### ISSUE TITLE
[HunyuanVideo-1.5-Diffusers-480p_t2v_distilled] pcc=nan in PCC gated E2E nightly test

### ISSUE THREAD

**Issue by kamalrajkannan78** [tenstorrent/tt-xla#5762]
### Problem description

- The PCC-gated nightly test of the [HunyuanVideo-1.5-Diffusers-480p_t2v_distilled](https://huggingface.co/hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v_distilled) e2e pipeline (transformer on TT) shows NaN PCC - [jul22_hunyuan_video_1_5_transformer_nan_pcc.log.zip](https://github.com/user-attachments/files/30292458/jul22_hunyuan_video_1_5_transformer_nan_pcc.log.zip), [pipeline.py](https://github.com/user-attachments/files/30292460/pipeline.py), [test_hunyuan_video_1_5_pipeline.py](https://github.com/user-attachments/files/30292459/test_hunyuan_video_1_5_pipeline.py)

### Root cause analysis

- Manual bisection experiments ([manual_bisect.zip](https://github.com/user-attachments/files/30294393/manual_bisect.zip)) reveal: 

| Iteration | PCC      |
|-----------|----------|
| 1         | 0.984375 |
| 2         | 0.022217 |
| 3         | nan      |

- Huge shift in PCC on the second iteration itself, so it should be fixed first.
- Further slicing reveals that the drop originates from this [LayerNorm](https://github.com/huggingface/diffusers/blob/a3608b512ed7248499a44c61d954965ed9bdae4d/src/diffusers/models/normalization.py#L169): PCC changed from 0.984375 to 0.91796875.
- Weights available at `kkannan/jul23_hunyuan_video1_5_pcc_test`.
- But unable to replicate the drop with a sanity check on a single chip — [jul22_hunyuan_video_ln.log](https://github.com/user-attachments/files/30293540/jul22_hunyuan_video_ln.log), [test_layernorm.py](https://github.com/user-attachments/files/30293541/test_layernorm.py). (Note: the model runs on MC, but elementwise_affine=False means that layer has no weights and is replicated across chips, so it should reproduce the drop on SC too.)
- This LayerNorm decomposition workaround (inspired by [PR#824](https://github.com/tenstorrent/tt-forge-models/pull/824)) solved the NaN PCC and boosted PCC to >0.90 across all runs:

```
def _decompose_layernorm(model):
"""Replace every nn.LayerNorm's forward with explicit mean/var/rsqrt ops in
the model dtype (bf16).


Empirical workaround for a TT PCC drop / NaN in the early transformer blocks
of the first DiT inference step, traced to LayerNorm but not reproducible in
a standalone sanity check (still under investigation). Details:
https://github.com/tenstorrent/tt-xla/issues/5762

Covers norm1/norm2's inner LayerNorm; NOT the QK-norm (rms_norm) in attention.
"""
 for mod in model.modules():
    if isinstance(mod, torch.nn.LayerNorm):

        def _fwd(x, m=mod):
            mu = x.mean(-1, keepdim=True)
            var = (x - mu).pow(2).mean(-1, keepdim=True)
            y = (x - mu) * torch.rsqrt(var + m.eps)
            if m.weight is not None:
                y = y * m.weight
            if m.bias is not None:
                y = y + m.bias
            return y

        mod.forward = _fwd
```
- Cross-verified the correctness of this hack with a sanity check across different cases; it achieved PCC>0.99 on all cases.
  - [jul22_ln_decompose_check.log](https://github.com/user-attachments/files/30293830/jul22_ln_decompose_check.log), [test_ln_check.py](https://github.com/user-attachments/files/30293864/test_ln_check.py)

### TODO

- [ ] Fix the PCC drop in the compiler and remove the workaround.

**Comment by kamalrajkannan78** [comment]
## 28/07/2026

### Root cause analysis

- While returning `hidden_states` and `encoder_hidden_states` from the transformer([quick_exp.zip](https://github.com/user-attachments/files/30473836/quick_exp.zip)), the following PCC values were observed across forward runs: 

| Run | hidden_states pcc | encoder_hidden_states pcc |
|-----|--------------------|-----------------------------|
| 1st | 0.984375           | -0.000000                   |
| 2nd | 0.047852           | nan                          |
| 3rd | nan                | nan                          |


- Since `encoder_hidden_states` from one forward run feeds into the next, `hidden_states` lags one run behind `encoder_hidden_states`' corruption: run 2's `hidden_states` is already low-PCC from run 1's bad `encoder_hidden_states`, and run 3's `hidden_states` turns NaN because run 2's `encoder_hidden_states` had already turned NaN. So the root cause sits in `encoder_hidden_states`, and that's what should be debugged first.
- Slicing further down, this is the minimal sanity needed to reproduce the drop - [test_initial_sanity.py](https://github.com/user-attachments/files/30474400/test_initial_sanity.py), [jul28_initial_sanity.log](https://github.com/user-attachments/files/30474399/jul28_initial_sanity.log)
- Since the sanity involves SDPA, I tried enabling the optimization level to check whether a fused SDPA at the ttnn level resolves the PCC drop, but it did not.

### Repro

```
git checkout kkannan/jul28_hunyuan_video1_5_nan_pcc_debug
git submodule update --init --recursive

# sanity run 
pytest tests/torch/models/Hunyuan_1_5/test_reduced_sanity.py -svv 
```

### Logs

- [manual_bisect.zip](https://github.com/user-attachments/files/30473835/manual_bisect.zip)
- [sanities.zip](https://github.com/user-attachments/files/30473834/sanities.zip) 

### Next steps

- Generate a repro using EmitPy, then reduce the sanity directly there.

**Comment by kamalrajkannan78** [comment]
## 29/07/2026

### Root cause: missing `torch.compile(backend="tt")`

- While restructuring the pipeline code from tt-xla into tt-forge-models, the `torch.compile(backend="tt")` call on the transformer was missed. Without it the DiT ran on the torch_xla lazy-tensor path instead of being compiled, which is what caused the PCC drop — in both the full model and the standalone sanity.
- After adding `torch.compile(backend="tt")` after the device move and shard annotations(matching the `hunyuan_image_2_1` pipeline):
   - **Sanity:** PCC = 1.000000 (was 0.012939)
   - **Pipeline (e2e, 10 denoising steps):** DiT PCC 0.910 → 0.996, above the 0.90 threshold at
  every iteration; VAE decode completes and the test passes
- Both results are with the `_decompose_layernorm` workaround **removed**, so the workaround is no longer needed and can be dropped.

### Logs

- [jul29_inital_sanity_fix_t.log](https://github.com/user-attachments/files/30513847/jul29_inital_sanity_fix_t.log)
- [test_initial_sanity.py](https://github.com/user-attachments/files/30513848/test_initial_sanity.py)
- [pcc_check.zip](https://github.com/user-attachments/files/30517943/pcc_check.zip)
